File: kid-problem-ui-test-selenium/KPUITest/Pages/AssignmentQueryPage.cs
using OpenQA.Selenium;

namespace KPUITest.Pages;

public sealed class AssignmentQueryPage : PageBase
{
    public readonly static string PageUrl = $"{Constants.BASE_URL}/assignments/all";

    private readonly By startTimeField = By.Name("startTime");
    private readonly By endTimeField = By.Name("endTime");
    private readonly By searchButton = By.CssSelector("app-assignment-list-view form button:nth-child(2)");
    private readonly By returnResultTableRows = By.CssSelector("app-assignment-list-view tr");
    private readonly By countField = By.CssSelector("app-assignment-list-view .w3-cell-row div.w3-cell:not(.w3-container)");    

    public AssignmentQueryPage(IWebDriver driver) : base(driver)
    { }

    public override bool IsPageLoaded(bool withData = false)
    {
        var loaded = _wait.Until(d => d.FindElement(By.CssSelector("app-assignment-query > h3")).Text.Contains("List Latest Assignments"));
        loaded = loaded && _wait.Until(d => d.FindElement(By.CssSelector("app-assignment-query mat-date-range-input")) != null);
        loaded = loaded && IsNotLoading();
        return loaded;
    }

    public void SetQueryDateRange(string startTimeText, string endTimeText)
    {
        var startTimeElement = _wait.Until(d => d.FindElement(startTimeField));
        if (startTimeElement != null)
        {
            startTimeElement.SendKeys(Keys.Control + "a");
            startTimeElement.SendKeys(Keys.Backspace);
            startTimeElement.SendKeys(startTimeText);
        }

        var endTimeElement = _wait.Until(d => d.FindElement(endTimeField));
        if (endTimeElement != null)
        {
            endTimeElement.SendKeys(Keys.Control + "a");
            endTimeElement.SendKeys(Keys.Backspace);
            endTimeElement.SendKeys(endTimeText);
        }
    }

    public void ClickSearchButton()
    {
        var btn = _wait.Until(d => d.FindElement(searchButton));
        if (btn != null)
        {
            btn.Click();
            IsNotLoading();
        }
    }

    public string GetAssignmentTitleFromQueryResult(int index)
    {
        var hasResult = _wait.Until(d => d.FindElements(returnResultTableRows).Count > 0);
        if (hasResult)
        {
            var td = _wait.Until(d => d.FindElement(By.CssSelector($"app-assignment-list-view tbody tr:nth-child({index + 1}) > td:nth-child(2)")));
            if (td != null)
            {
                return td.Text;
            }
        }
        return "";
    }

    public AssignmentViewPage? ClickAssignmentTitleInQueryResults(int index)
    {
        var hasResult = _wait.Until(d => d.FindElements(returnResultTableRows).Count > 0);
        if (hasResult)
        {
            var link = _wait.Until(d => d.FindElement(By.CssSelector($"app-assignment-list-view tbody tr:nth-child({index + 1}) td:nth-child(2) a")));
            if (link != null)
            {
                link.Click();
                if (_wait.Until(d => d.Url.Contains("/assignment/view/")))
                {
                    return new AssignmentViewPage(_driver);
                }
            }
        }
        return null;
    }

    public int GetCountOfQueryResults()
    {
        var hasResult = _wait.Until(d => d.FindElements(returnResultTableRows).Count > 0);
        if (hasResult)
        {
            var countElement = _wait.Until(d => d.FindElement(countField));
            if (countElement != null)
            {
                var countText = countElement.Text.Replace("Count of all records:", "").Trim();
                if (string.IsNullOrWhiteSpace(countText))
                {
                    return 0;
                }
                else
                {
                    if (int.TryParse(countText, out int count))
                    {
                        return count;
                    }
                }
            }
        }

        return 0;
    }

    private bool IsNotLoading()
    {
        By progressBar = By.CssSelector("mat-progress-bar");
        return _wait.Until(d => d.FindElements(progressBar).Count == 0);
    }

}

File: kid-problem-ui-test-playwright/pages/AssignmentQueryPage.ts
import { expect, Locator, Page } from "@playwright/test";
import { BASE_URL } from "../utils/constants";
import { waitForNoMatProgressBar } from "../utils/waitForNoMatProgressBar";

export class AssignmentQueryPage {
    readonly page: Page;

    constructor(page: Page) {
        this.page = page;
    }

    async goTo() {
        await this.page.goto(`${BASE_URL}/assignments/all`);
        await waitForNoMatProgressBar(this.page);
    }

    async query(startTime: string | null, endTime: string | null) {
        if (startTime !== null) {
            await this.page.getByPlaceholder('Start Date').fill(startTime);
        }

        if (endTime !== null) {
            await this.page.getByPlaceholder('End Date').fill(endTime);
        }

        await this.page.locator('app-assignment-list-view form button:nth-child(2)').click();
    }

    getAssignmentTitleFromQueryResult(index: number): Locator {
        return this.page.locator(`app-assignment-list-view tbody tr:nth-child(${index + 1}) > td:nth-child(2)`);
    }

    async clickAssignmentTitleInQueryResults(index: number) {
        const link = this.page.locator(`app-assignment-list-view tbody tr:nth-child(${index + 1}) td:nth-child(2) a`);
        await expect(link).toBeVisible();
        await link.click();
        await waitForNoMatProgressBar(this.page);
    }

    async clickDoAssignmentInQueryResults(index: number) {
        const btn = this.page.locator(`app-assignment-list-view tbody tr:nth-child(${index + 1}) button`);
        await expect(btn).toBeVisible();
        await btn.click();
        await waitForNoMatProgressBar(this.page);
    }

    getCountOfQueryResultsField(): Locator {
        const countField = this.page.locator('app-assignment-list-view .w3-cell-row div.w3-cell:not(.w3-container)');
        return countField;
    }
}